### PR TITLE
support specifying struct fields to inject

### PR DESCRIPTION
**Pull request by shantuo**
This is a  breaking change. There will be follow-up change for updating guides.

Updates #36

### CODE REVIEW

File: internal/wire/parse.go [hunk 1]
```diff
@@ -560,13 +566,6 @@ func (oc *objectCache) processExpr(info *types.Info, pkgPath string, expr ast.Ex
 			return nil, []error{notePosition(exprPos, errors.New("unknown pattern"))}
 		}
 	}
-	if tn := structArgType(info, expr); tn != nil {
```
[the remark this answers is not in the document]
Comment (shantuo) [reply internal/wire/parse.go]: Done.

File: wire.go [hunk 2]
```diff
@@ -137,6 +136,26 @@ func InterfaceValue(typ interface{}, x interface{}) ProvidedValue {
 	return ProvidedValue{}
 }
 
+// A StructProvider represents a named struct.
+type StructProvider struct{}
+
+// Struct specifies that the given struct type will be provided by filling in the fields
+// in the struct that have the names given. Each of the arguments must be a name
+// to the field they wish to reference.
```
[the remark this answers is not in the document]
Comment (shantuo) [reply wire.go]: Done.

**Comment by zombiezen** [comment]
One idea that came to mind: while we don't have version information, we could introspect the contents of the `github.com/google/wire` package (the one that is making the `wire.NewSet` call). Based on that, we could allow the old functionality if `wire.Struct` is present, and forbid it if `wire.Struct` is present. This would give us a natural way of avoiding breakage until the user of Wire explicitly upgrades the package.

**Comment by shantuo** [comment]
Discussed offline, the plan is to deprecate the old way instead of removing it. Add a warning message when the struct literal is processed.

File: internal/wire/parse.go [hunk 3]
```diff
@@ -732,6 +739,43 @@ func funcOutput(sig *types.Signature) (outputSignature, error) {
 	}
 }
 
+// processStructLitProvider creates a provider for a named struct type.
+// It produces pointer and non-pointer variants via two values in Out.
+//
+// This is a copy of the old processStructProvider, which is deprecated now.
+func processStructLitProvider(fset *token.FileSet, typeName *types.TypeName) (*Provider, []error) {
```
Comment (zombiezen) [inline internal/wire/parse.go]: nit: s/Lit/Literal/ for clarity.

File: internal/wire/parse.go [hunk 4]
```diff
@@ -732,6 +739,43 @@ func funcOutput(sig *types.Signature) (outputSignature, error) {
...
+	fmt.Println("Deprecated: use wire.Struct to inject a struct instead,",
```
Comment (zombiezen) [inline internal/wire/parse.go]: Two things here:

1.  A deprecation notice should go to stderr to stay in the same stream as the other error messages.
2.  This deprecation message does not include enough information for the user to find the element that needs upgrading. Let's include at least the type name, and preferably the file/line number.

File: wire.go [hunk 5]
```diff
@@ -47,6 +47,7 @@ type ProviderSet struct{}
 // Passing a struct value of type S to NewSet declares that both S and *S will
 // be provided by creating a new value of the appropriate type by filling in
 // each field of S using the provider of the field's type.
+// Deprecated: use wire.Struct.
```
Comment (zombiezen) [inline wire.go]: Let's deemphasize this paragraph by moving it to the end of the doc comment.

Also, I think we could frame the deprecation to give the reader more information with slightly different wording. I propose:

**For compatibility with older versions of Wire,** passing a struct value of type S to NewSet declares that both S and *S will be provided by creating a new value of the appropriate type by filling in each field of S using the provider of the field's type. **This form is deprecated and will be removed in a future version of Wire: new providers sets should use wire.Struct.**

File: wire.go [hunk 6]
```diff
@@ -141,7 +143,8 @@ type StructProvider struct{}
 
 // Struct specifies that the given struct type will be provided by filling in the fields
 // in the struct that have the names given. Each of the arguments must be a name
-// to the field they wish to reference.
+// to the field they wish to reference. As a special case, if a single name "*"
+// is given, then all of the fields in the struct will be filled in.
```
Comment (zombiezen) [inline wire.go]: Have we added docs about this to the user guide? If not, let's open an issue.

File: internal/wire/parse.go (hunk 4 above)
Comment (shantuo) [reply internal/wire/parse.go]: Done.

File: wire.go (hunk 5 above)
Comment (shantuo) [reply wire.go]: Done.

**Pull request merged**